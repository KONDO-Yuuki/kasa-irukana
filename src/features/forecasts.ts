import {createAsyncThunk, createSlice} from '@reduxjs/toolkit';
// import {forecastsApi} from '../api/forecasts';

export type IsUmbrellaNecessary = 'NECESSARY' | 'MAYBE' | 'UNNECESSARY';

export type Forecast = {
  date: string;
  dateLabel: string;
  telop: string;
  detail: {weather: string; wind: string};
  temperature: {min: string | null; max: string | null};
  chanceOfRain: {
    T00_06: number | null;
    T06_12: number | null;
    T12_18: number | null;
    T18_24: number;
  };
  imageUrl: string;
};

export type ForecastsState = {
  values: Forecast[];
};

const initialState: ForecastsState = {
  values: [],
};

export const fetchForecastByCityCode = createAsyncThunk<any, string>(
  'forecasts/fetchByCityCode/',
  async citiyId => {
    const result = await fetch(
      `https://weather.tsukumijima.net/api/forecast/city/${citiyId}`,
    );
    return result.json();
  },
);

export const forecastsSlice = createSlice({
  name: 'forecasts',
  initialState: initialState,
  reducers: {},
  extraReducers: builder => {
    builder.addCase(fetchForecastByCityCode.fulfilled, (state, action) => {
      state.values = castApiResult(action.payload);
    });
  },
});

export const castApiResult = (body: any): Forecast[] => {
  const validBody = validate(body);
  return validBody.forecasts.map((forecast: any) => {
    return {
      date: forecast.date,
      dateLabel: forecast.dateLabel,
      telop: forecast.telop,
      detail: {
        weather: forecast.detail.weather,
        wind: forecast.detail.wind,
      },
      temperature: {
        min: forecast.temperature.min.celsius,
        max: forecast.temperature.max.celsius,
      },
      chanceOfRain: {
        T00_06: parseChanceOfRain(forecast.chanceOfRain.T00_06),
        T06_12: parseChanceOfRain(forecast.chanceOfRain.T06_12),
        T12_18: parseChanceOfRain(forecast.chanceOfRain.T12_18),
        T18_24: parseChanceOfRain(forecast.chanceOfRain.T18_24),
      },
      imageUrl: forecast.image.url,
    };
  });
};

const validate = (body: any): any => {
  // todo 実装
  // joiなどでapiのshapeが壊れていないことをバリデートする
  return body;
};

/**
 * 数字 + '%' のstring型で定義されるchanceOfRainのパース
 * @param rainLabel
 */
const parseChanceOfRain = (rainLabel: string) => {
  // 1桁~3桁の数字(0 ~ 100)のみmatch
  const matched = rainLabel.match(/[0-9]{1,3}/);
  if (!matched) {
    return null;
  }
  return Number(matched);
};

export default forecastsSlice.reducer;
