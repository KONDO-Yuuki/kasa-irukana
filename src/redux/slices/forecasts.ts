import {createAsyncThunk, createSlice} from '@reduxjs/toolkit';
import {
  Forecast,
  ForecastsState,
  UmbrellaNecessary,
  UmbrellaNecessaryState,
} from '../../types/forecasts';

const initialState: ForecastsState = {
  startForecasts: [],
  goalForecasts: [],
  umbrellaNecessaryStates: [],
};

const createFetchForecastByCityCode = (name: string) =>
  createAsyncThunk<any, string>(name, async citiyId => {
    const result = await fetch(
      `https://weather.tsukumijima.net/api/forecast/city/${citiyId}`,
    );
    return result.json();
    // todo エラーハンドリングの実装
  });
export const fetchStartForecastByCityCode = createFetchForecastByCityCode(
  'fetchStartForecastByCityCode',
);
export const fetchEndForecastByCityCode = createFetchForecastByCityCode(
  'fetchEndForecastByCityCode',
);

export const forecastsSlice = createSlice({
  name: 'forecasts',
  initialState: initialState,
  reducers: {},
  extraReducers: builder => {
    builder.addCase(fetchStartForecastByCityCode.fulfilled, (state, action) => {
      state.startForecasts = castApiResult(action.payload);
      if (
        state.startForecasts.length === 3 &&
        state.goalForecasts.length === 3
      ) {
        state.umbrellaNecessaryStates = calcUmbrellaNecessaries(
          state.startForecasts,
          state.goalForecasts,
        );
      }
    });
    builder.addCase(fetchEndForecastByCityCode.fulfilled, (state, action) => {
      state.goalForecasts = castApiResult(action.payload);
      if (
        state.startForecasts.length === 3 &&
        state.goalForecasts.length === 3
      ) {
        state.umbrellaNecessaryStates = calcUmbrellaNecessaries(
          state.startForecasts,
          state.goalForecasts,
        );
      }
    });
  },
});

export const castApiResult = (body: any): Forecast[] => {
  const validBody = validate(body);
  return validBody.forecasts.map((forecast: any) => {
    return {
      date: forecast.date,
      dateLabel: forecast.dateLabel,
      telop: forecast.telop,
      position: validBody.location.city,
      title: validBody.title,
      detail: {
        weather: forecast.detail.weather,
        wind: forecast.detail.wind,
      },
      temperature: {
        min: forecast.temperature.min.celsius,
        max: forecast.temperature.max.celsius,
      },
      chanceOfRainBy6Hours: [
        parseChanceOfRain(forecast.chanceOfRain.T00_06),
        parseChanceOfRain(forecast.chanceOfRain.T06_12),
        parseChanceOfRain(forecast.chanceOfRain.T12_18),
        parseChanceOfRain(forecast.chanceOfRain.T18_24),
      ],
      imageUrl: forecast.image.url,
    };
  });
};

const validate = (body: any): any => {
  // todo 実装
  // joiなどでapiのshapeが壊れていないことをバリデートする
  return body;
};

/**
 * 数字 + '%' のstring型で定義されるchanceOfRainのパース
 * @param rainLabel
 */
const parseChanceOfRain = (rainLabel: string) => {
  // 1桁~3桁の数字(0 ~ 100)のみmatch
  const matched = rainLabel.match(/[0-9]{1,3}/);
  if (!matched) {
    return null;
  }
  return Number(matched);
};

/**
 *
 * @param startForecasts 出発地のForecast
 * @param goalForecasts 到着地のForecast
 * @returns UmbrellaNecessaryモデル(日付と傘が必要かどうかのステータス)
 */
export const calcUmbrellaNecessaries = (
  startForecasts: Forecast[],
  goalForecasts: Forecast[],
): UmbrellaNecessary[] => {
  return startForecasts.map(sf => {
    const gf = goalForecasts.find(g => g.date === sf.date);
    if (!gf) {
      return {date: sf.date, label: 'UNKNOWN'};
    }
    return {date: sf.date, label: isAmbrellaNecessary(sf, gf)};
  });
};

export const isAmbrellaNecessary = (
  startForecast: Forecast,
  goalForecast: Forecast,
): UmbrellaNecessaryState => {
  return startForecast.chanceOfRainBy6Hours.reduce(
    (
      accum: UmbrellaNecessaryState,
      currentStartPercent: number | null,
      i: number,
    ) => {
      const currentGoalPercent = goalForecast.chanceOfRainBy6Hours[i];
      // 0 : numberの事を考えてnullチェック
      if (currentStartPercent === null || currentGoalPercent === null) {
        // null(すでに経過した時刻)の場合は無視する
        return accum;
      }

      if (i === 0) {
        // 0 ~ 6時の天気は加味しない
        return accum;
      }

      if (
        accum === 'NECESSARY' ||
        currentStartPercent >= 30 ||
        currentGoalPercent >= 30
      ) {
        // すでにNECESSARY判定 or 降水確率30％より多いとNECESSARY
        return 'NECESSARY';
      }

      if (
        accum === 'MAYBE' ||
        currentStartPercent > 0 ||
        currentGoalPercent > 0
      ) {
        // すでにMAYBE判定 or 降水確率が0％より多いとMAYBE
        return 'MAYBE';
      }

      if (
        accum === 'UNNECESSARY' ||
        (currentStartPercent === 0 && currentGoalPercent === 0)
      ) {
        // 降水確率0％のときのみUNNECESSARY
        return 'UNNECESSARY';
      }

      return accum;
    },
    'UNKNOWN',
  );
};

export default forecastsSlice.reducer;
