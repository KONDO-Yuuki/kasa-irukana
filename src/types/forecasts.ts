export type UmbrellaNecessaryState =
  | 'NECESSARY'
  | 'MAYBE'
  | 'UNNECESSARY'
  | 'UNKNOWN';

export type UmbrellaNecessary = {
  date: string;
  label: UmbrellaNecessaryState;
};

export type Forecast = {
  date: string;
  position: string;
  title: string;
  dateLabel: string;
  telop: string;
  detail: {weather: string; wind: string};
  temperature: {min: string | null; max: string | null};
  chanceOfRainBy6Hours: Array<number | null>;
  imageUrl: string;
};

export type ForecastsState = {
  startForecasts: Forecast[];
  goalForecasts: Forecast[];
  umbrellaNecessaryStates: UmbrellaNecessary[];
  error: string | null | undefined;
};
export type Position = 'start' | 'goal';
