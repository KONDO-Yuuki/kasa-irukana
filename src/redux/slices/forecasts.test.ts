import 'react-native';
import {
  forecastsSample,
  forecastRowData,
  umbrelleNecessarySample,
} from '../../fixtures/forecasts';
import {UmbrellaNecessaryState} from '../../types/forecasts';
import {
  calcUmbrellaNecessaries,
  castApiResult,
  isAmbrellaNecessary,
  initialState,
  forecastsSlice,
  fetchStartForecastByCityCode,
  fetchGoalForecastByCityCode,
} from './forecasts';

describe('forecastsSlice', () => {
  const rejectedPayload = {
    errorMessage: '通信に失敗しました',
    errorDetail: 'result.data.error',
  };
  describe('init', () => {
    it('reinitialize', () => {
      const startFulfilledState = forecastsSlice.reducer(initialState, {
        type: fetchStartForecastByCityCode.fulfilled.type,
        payload: forecastRowData,
      });
      const state = forecastsSlice.reducer(
        startFulfilledState,
        forecastsSlice.actions.init,
      );
      expect(state).toEqual(initialState);
    });
  });
  describe('fetchStartForecastByCityCode', () => {
    it('fulfilled', () => {
      const action = {
        type: fetchStartForecastByCityCode.fulfilled.type,
        payload: forecastRowData,
      };
      const state = forecastsSlice.reducer(initialState, action);
      expect(state.startForecasts).toMatchSnapshot();
      expect(state.goalForecasts).toHaveLength(0);
      expect(state.umbrellaNecessaryStates).toHaveLength(0);
      expect(state.error).toBeFalsy();
    });
    it('rejected', () => {
      const action = {
        type: fetchStartForecastByCityCode.rejected.type,
        payload: rejectedPayload,
      };
      const state = forecastsSlice.reducer(initialState, action);
      expect(state.startForecasts).toHaveLength(0);
      expect(state.goalForecasts).toHaveLength(0);
      expect(state.error).toBe(rejectedPayload.errorMessage);
    });
  });
  describe('fetchGoalForecastByCityCode', () => {
    it('fulfilled', () => {
      const action = {
        type: fetchGoalForecastByCityCode.fulfilled.type,
        payload: forecastRowData,
      };
      const state = forecastsSlice.reducer(initialState, action);
      expect(state.startForecasts).toHaveLength(0);
      expect(state.goalForecasts).toMatchSnapshot();
      expect(state.umbrellaNecessaryStates).toHaveLength(0);
      expect(state.error).toBeFalsy();
    });
    it('rejected', () => {
      const action = {
        type: fetchGoalForecastByCityCode.rejected.type,
        payload: rejectedPayload,
      };
      const state = forecastsSlice.reducer(initialState, action);
      expect(state.startForecasts).toHaveLength(0);
      expect(state.goalForecasts).toHaveLength(0);
      expect(state.error).toBe(rejectedPayload.errorMessage);
    });
  });
  describe('umbrellaNecessaryStates', () => {
    it('fulfilled start and goal', () => {
      const startFulfilledState = forecastsSlice.reducer(initialState, {
        type: fetchStartForecastByCityCode.fulfilled.type,
        payload: forecastRowData,
      });
      const state = forecastsSlice.reducer(startFulfilledState, {
        type: fetchGoalForecastByCityCode.fulfilled.type,
        payload: forecastRowData,
      });
      expect(state.umbrellaNecessaryStates).toHaveLength(3);
      expect(state.umbrellaNecessaryStates).toMatchSnapshot();
      expect(state.error).toBeFalsy();
    });
    it('fulfilled start but rejected goal', () => {
      const startFulfilledState = forecastsSlice.reducer(initialState, {
        type: fetchStartForecastByCityCode.fulfilled.type,
        payload: forecastRowData,
      });
      const state = forecastsSlice.reducer(startFulfilledState, {
        type: fetchGoalForecastByCityCode.rejected.type,
        payload: rejectedPayload,
      });
      expect(state.umbrellaNecessaryStates).toHaveLength(0);
      expect(state.error).toBe(rejectedPayload.errorMessage);
    });
    it('rejected start but fulfilled goal', () => {
      const startRejectedState = forecastsSlice.reducer(initialState, {
        type: fetchStartForecastByCityCode.rejected.type,
        payload: rejectedPayload,
      });
      const state = forecastsSlice.reducer(startRejectedState, {
        type: fetchGoalForecastByCityCode.fulfilled.type,
        payload: forecastRowData,
      });
      expect(state.umbrellaNecessaryStates).toHaveLength(0);
      expect(state.error).toBe(rejectedPayload.errorMessage);
    });
    it('rejected start and goal', () => {
      const startRejectedState = forecastsSlice.reducer(initialState, {
        type: fetchStartForecastByCityCode.rejected.type,
        payload: rejectedPayload,
      });
      const state = forecastsSlice.reducer(startRejectedState, {
        type: fetchGoalForecastByCityCode.rejected.type,
        payload: rejectedPayload,
      });
      expect(state.umbrellaNecessaryStates).toHaveLength(0);
      expect(state.error).toBe(rejectedPayload.errorMessage);
    });
  });
});

describe('castApiResult', () => {
  it('fixture同士が変換できること', () => {
    const result = castApiResult(forecastRowData);
    expect(result).toMatchObject(forecastsSample);
  });
});

describe('calcUmbrellaNecessaries', () => {
  it('startForecastsに対応したgoalForecastsが無いとUNKNOWNになること', () => {
    const result = calcUmbrellaNecessaries(forecastsSample, []);
    expect(result).toMatchObject([
      {date: '2022-04-24', label: 'UNKNOWN'},
      {date: '2022-04-25', label: 'UNKNOWN'},
      {date: '2022-04-26', label: 'UNKNOWN'},
    ]);
  });
  it('fixture同士が変換できること', () => {
    const result = calcUmbrellaNecessaries(forecastsSample, forecastsSample);
    expect(result).toMatchObject(umbrelleNecessarySample);
  });
});
test.each<
  [
    number | null,
    number | null,
    number | null,
    number | null,
    number | null,
    number | null,
    number | null,
    number | null,
    UmbrellaNecessaryState,
  ]
>([
  [100, 100, 100, 100, 100, 100, 100, 100, 'NECESSARY'],
  [null, 20, 0, 0, null, 100, 0, 0, 'NECESSARY'],
  [0, 0, 0, 0, 0, 0, 0, 0, 'UNNECESSARY'],
  [null, 0, 0, 0, null, 0, 0, 0, 'UNNECESSARY'],
  [null, 20, 0, 0, null, 0, 0, 0, 'MAYBE'],

  [null, 30, 0, 0, null, 0, 0, 0, 'NECESSARY'], // 境界値
  [null, 29, 0, 0, null, 0, 0, 0, 'MAYBE'], // 境界値
  [null, 1, 0, 0, null, 0, 0, 0, 'MAYBE'], // 境界値
  [null, 0, 0, 0, null, 0, 0, 0, 'UNNECESSARY'], // 境界値

  [null, null, 0, 0, null, null, 0, 0, 'UNNECESSARY'],
  [null, null, null, 0, null, null, null, 100, 'NECESSARY'],
  [100, 0, 0, 0, 100, 0, 0, 0, 'UNNECESSARY'], // 明け方雨でも傘はいらない
  [100, 30, 0, 0, 100, 0, 0, 0, 'NECESSARY'],
])(
  'isAmbrellaNecessary関数　start側の T00_06:%s, T06_12:%s、ST12_18:%s、T18_24:%s goal側の T00_06:%s, T06_12:%s、ST12_18:%s、T18_24:%sだと%sになる',
  (
    ST00_06,
    ST06_12,
    ST12_18,
    ST18_24,
    GT00_06,
    GT06_12,
    GT12_18,
    GT18_24,
    exp,
  ) => {
    const result = isAmbrellaNecessary(
      {
        ...forecastsSample[0],
        chanceOfRainBy6Hours: [ST00_06, ST06_12, ST12_18, ST18_24],
      },

      {
        ...forecastsSample[0],
        chanceOfRainBy6Hours: [GT00_06, GT06_12, GT12_18, GT18_24],
      },
    );
    expect(result).toBe(exp);
  },
);
