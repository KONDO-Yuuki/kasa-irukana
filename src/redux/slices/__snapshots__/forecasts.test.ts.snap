// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forecastsSlice fetchGoalForecastByCityCode fulfilled 1`] = `
Array [
  Object {
    "chanceOfRainBy6Hours": Array [
      null,
      null,
      null,
      60,
    ],
    "date": "2022-04-24",
    "dateLabel": "今日",
    "detail": Object {
      "weather": "雨　夜遅く　くもり",
      "wind": "北の風",
    },
    "imageUrl": "https://www.jma.go.jp/bosai/forecast/img/313.svg",
    "position": "東京",
    "telop": "雨のち曇",
    "temperature": Object {
      "max": null,
      "min": null,
    },
    "title": "東京都 東京 の天気",
  },
  Object {
    "chanceOfRainBy6Hours": Array [
      0,
      0,
      0,
      0,
    ],
    "date": "2022-04-25",
    "dateLabel": "明日",
    "detail": Object {
      "weather": "晴れ　時々　くもり",
      "wind": "北の風　後　南の風",
    },
    "imageUrl": "https://www.jma.go.jp/bosai/forecast/img/101.svg",
    "position": "東京",
    "telop": "晴時々曇",
    "temperature": Object {
      "max": "27",
      "min": "14",
    },
    "title": "東京都 東京 の天気",
  },
  Object {
    "chanceOfRainBy6Hours": Array [
      20,
      20,
      20,
      20,
    ],
    "date": "2022-04-26",
    "dateLabel": "明後日",
    "detail": Object {
      "weather": "くもり",
      "wind": "南の風　後　強く",
    },
    "imageUrl": "https://www.jma.go.jp/bosai/forecast/img/200.svg",
    "position": "東京",
    "telop": "曇り",
    "temperature": Object {
      "max": "25",
      "min": "15",
    },
    "title": "東京都 東京 の天気",
  },
]
`;

exports[`forecastsSlice fetchStartForecastByCityCode fulfilled 1`] = `
Array [
  Object {
    "chanceOfRainBy6Hours": Array [
      null,
      null,
      null,
      60,
    ],
    "date": "2022-04-24",
    "dateLabel": "今日",
    "detail": Object {
      "weather": "雨　夜遅く　くもり",
      "wind": "北の風",
    },
    "imageUrl": "https://www.jma.go.jp/bosai/forecast/img/313.svg",
    "position": "東京",
    "telop": "雨のち曇",
    "temperature": Object {
      "max": null,
      "min": null,
    },
    "title": "東京都 東京 の天気",
  },
  Object {
    "chanceOfRainBy6Hours": Array [
      0,
      0,
      0,
      0,
    ],
    "date": "2022-04-25",
    "dateLabel": "明日",
    "detail": Object {
      "weather": "晴れ　時々　くもり",
      "wind": "北の風　後　南の風",
    },
    "imageUrl": "https://www.jma.go.jp/bosai/forecast/img/101.svg",
    "position": "東京",
    "telop": "晴時々曇",
    "temperature": Object {
      "max": "27",
      "min": "14",
    },
    "title": "東京都 東京 の天気",
  },
  Object {
    "chanceOfRainBy6Hours": Array [
      20,
      20,
      20,
      20,
    ],
    "date": "2022-04-26",
    "dateLabel": "明後日",
    "detail": Object {
      "weather": "くもり",
      "wind": "南の風　後　強く",
    },
    "imageUrl": "https://www.jma.go.jp/bosai/forecast/img/200.svg",
    "position": "東京",
    "telop": "曇り",
    "temperature": Object {
      "max": "25",
      "min": "15",
    },
    "title": "東京都 東京 の天気",
  },
]
`;

exports[`forecastsSlice umbrellaNecessaryStates fulfilled start and goal 1`] = `
Array [
  Object {
    "date": "2022-04-24",
    "label": "NECESSARY",
  },
  Object {
    "date": "2022-04-25",
    "label": "UNNECESSARY",
  },
  Object {
    "date": "2022-04-26",
    "label": "MAYBE",
  },
]
`;
